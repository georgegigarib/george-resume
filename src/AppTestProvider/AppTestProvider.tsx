import React from 'react'
import { I18nextProvider } from 'react-i18next'
import i18n from '@/locales/i18n.config'
import store from '@/store'
import { Provider } from 'react-redux'
import { BrowserRouter as Router, Routes, Route } from 'react-router-dom'

const AppTestProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <I18nextProvider i18n={i18n}>
      <Provider store={store}>
        <Router>
          <Routes>
            <Route path="/" element={children} />
          </Routes>
        </Router>
      </Provider>
    </I18nextProvider>
  )
}

export default AppTestProvider
