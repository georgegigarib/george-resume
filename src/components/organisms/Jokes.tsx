import React, { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import IconButton from '@mui/material/IconButton'
import CachedIcon from '@mui/icons-material/Cached'

export default function Jokes(): React.ReactElement {
  const { t } = useTranslation()
  const jokes = [t('me.jokes.joke1'), t('me.jokes.joke2'), t('me.jokes.joke3'), t('me.jokes.joke4')]

  const [currentJoke, setCurrentJoke] = useState(0)
  const [fade, setFade] = useState(true)

  useEffect(() => {
    const interval = setInterval(() => {
      handleChangeJoke()
    }, 8000)

    return () => clearInterval(interval)
  }, [currentJoke])

  const handleChangeJoke = () => {
    setFade(false)
    setTimeout(() => {
      let randomIndex
      do {
        randomIndex = Math.floor(Math.random() * jokes.length)
      } while (randomIndex === currentJoke)

      setCurrentJoke(randomIndex)
      setFade(true)
    }, 500)
  }

  return (
    <div className="flex relative flex-col justify-between w-full h-full p-4 overflow-hidden rounded-xl border-[2px] bg-app-lightModeBoxes-box16 dark:bg-app-darkModeBoxes-box16 border-app-darkModeBoxes-box16 dark:border-app-lightModeBoxes-box16">
      <div className={`transition-opacity duration-500 ${fade ? 'opacity-100' : 'opacity-0'} h-full`}>
        <p data-testid="current-joke" className="text-gray-700 dark:text-gray-100 h-full">
          {jokes[currentJoke]}
        </p>
      </div>

      <IconButton
        aria-label="refresh joke"
        data-testid="refresh-joke-button"
        onClick={handleChangeJoke}
        sx={{
          position: 'absolute',
          bottom: '0px',
          right: '0px',
          color: 'inherit',
        }}
      >
        <CachedIcon fontSize="medium" />
      </IconButton>
    </div>
  )
}
