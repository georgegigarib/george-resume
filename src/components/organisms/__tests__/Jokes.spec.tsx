import { render, screen, fireEvent, act } from '@testing-library/react'
import Jokes from '@/components/organisms/Jokes'
import { vi, describe, afterEach, it, expect, beforeEach } from 'vitest'
import AppTestProvider from '@/AppTestProvider/AppTestProvider'

describe('Jokes Component', () => {
  const renderComponent = () =>
    render(
      <AppTestProvider>
        <Jokes />
      </AppTestProvider>
    )

  beforeEach(() => {
    vi.useFakeTimers()
  })

  afterEach(() => {
    vi.clearAllTimers()
  })

  it('matches the snapshot', () => {
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })

  it('changes the joke after the interval', () => {
    renderComponent()

    const jokeElement = screen.getByTestId('current-joke')
    const firstJoke = jokeElement.textContent

    expect(firstJoke).toBeTruthy()

    act(() => {
      vi.advanceTimersByTime(10000) // Simulate 10 seconds passing
    })

    const updatedJokeElement = screen.getByTestId('current-joke')
    const newJoke = updatedJokeElement.textContent
    expect(newJoke).not.toBe(firstJoke)
  })

  it('changes the joke when refresh button is clicked', () => {
    renderComponent()

    const jokeElement = screen.getByTestId('current-joke')
    const firstJoke = jokeElement.textContent

    expect(firstJoke).toBeTruthy()

    const refreshButton = screen.getByTestId('refresh-joke-button')
    fireEvent.click(refreshButton)

    act(() => {
      vi.advanceTimersByTime(2000) // Simulate time for transition
    })

    const updatedJokeElement = screen.getByTestId('current-joke')
    const newJoke = updatedJokeElement.textContent
    expect(newJoke).not.toBe(firstJoke)
  })
})
