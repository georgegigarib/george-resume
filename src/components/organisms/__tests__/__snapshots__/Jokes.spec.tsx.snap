// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Jokes Component > matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex relative flex-col justify-between w-full h-full p-4 overflow-hidden rounded-xl bg-gray-200 dark:bg-blue-900"
  >
    <div
      class="transition-opacity duration-500 opacity-100 h-full"
    >
      <p
        class="text-gray-700 dark:text-gray-100 h-full"
        data-testid="current-joke"
      >
        I decided to create this presentation because, let's face it, my jokes arenâ€™t as good as my code... but here we are.
      </p>
    </div>
    <button
      aria-label="refresh joke"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-5fjsz0-MuiButtonBase-root-MuiIconButton-root"
      data-testid="refresh-joke-button"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-20bmp1-MuiSvgIcon-root"
        data-testid="CachedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m19 8-4 4h3c0 3.31-2.69 6-6 6-1.01 0-1.97-.25-2.8-.7l-1.46 1.46C8.97 19.54 10.43 20 12 20c4.42 0 8-3.58 8-8h3zM6 12c0-3.31 2.69-6 6-6 1.01 0 1.97.25 2.8.7l1.46-1.46C15.03 4.46 13.57 4 12 4c-4.42 0-8 3.58-8 8H1l4 4 4-4z"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
